### PR TITLE
New Documentation

### PR DESCRIPTION
**Pull request by JackTemaki**
This is my proposal for the new documentation structure and content. Of course many parts are still missing or incomplete, but this should already be more comfortable for newcomers. Especially for the documentation of the internals (like how to write new Datasets or Layers, how to work with the search) I did not add anything yet.

### CODE REVIEW

File: docs/generate_units.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+from crnn.TFNetworkRecLayer import RecLayer
+from crnn.TFNativeOp import RecSeqCellOp
```
Comment (albertz) [inline docs/generate_units.py]: Better use the new-style import here: `from returnn. TFNativeOp import ...`.

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 2]
```diff
@@ -27,58 +27,26 @@ as well as some training parameters.
...
 use_tensorflow
     If you set this to ``True``, TensorFlow will be used.
 
-train / dev
+train / dev / eval
     The datasets. This can be a filename to a hdf-file.
     Or it can be a dict with an entry ``class`` where you can choose a from a variety
     of other dataset implementations, including many synthetic generated data.
+    ``train`` and ``dev`` are used during training, while ``eval`` is usually used to define the dataset for the
+    ``forward`` or ``search`` task.
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst]: Not quite correct. In general, I would not set `eval` inside the config. It would also be used during training then (in contrast to what you write), and not sure if this is what people expect. For the `"search"` task, I would esp recommend to use the `"search_data"` option instead.

File: docs/getting_started/basic_usage/data.rst [hunk 3]
```diff
@@ -0,0 +1,55 @@
+.. _data:
+
+=================
+Data Input/Output
+=================
+
+The parameters that are used to correctly define the data inputs are the three dataset variables ``train``, ``dev`` and
```
Comment (albertz) [inline docs/getting_started/basic_usage/data.rst]: Make newlines at sensible places (e.g. after a sentence, or after a logical break).
It should be such that further edits lead to minimal diffs.

File: docs/getting_started/basic_usage/data.rst [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    train = {'class': 'HDFDataset', 'files': ['path/to/training_data.hdf']}
+
+Most datasets follow the convention that the input data is sequential and has the label "``data``", and the target data
+is sparse and has the label "``classes``". In the case of the hdf file it could be that the input data are
+100-dimensional MFCCs and the target data are 5,000 word classes. Then, ``extern_data`` would have to be defined as:
+
+.. code-block::
+
+    extern_data = {'data': (100, 2), 'classes': (5000, 1)}
```
Comment (albertz) [inline docs/getting_started/basic_usage/data.rst]: I'm not sure that's a good initial example for `extern_data`. That is the old format for `num_outputs`, which I generally would not recommend to use.

File: docs/getting_started/basic_usage/network.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    "lstm1_bw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": -1, "from" : ["lstm0_fw", "lstm0_bw"] },
+
+    "lstm2_fw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": 1, "from" : ["lstm1_fw", "lstm1_bw"] },
+    "lstm2_bw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": -1, "from" : ["lstm1_fw", "lstm1_bw"] },
+
+    "out
+
+
+Defining Layers
+-------------------
```
Comment (albertz) [inline docs/getting_started/basic_usage/network.rst]: It's not a problem if the length of the `---` does not match?

File: docs/index.rst [hunk 6]
```diff
@@ -49,32 +49,53 @@ Questions can also be asked on
...
+    applications/lm.rst
+    applications/mt.rst
 
 .. toctree::
-    :maxdepth: 2
-    :glob:
+    :hidden:
+    :caption: Internals
 
+    internals/tech_overview.rst
```
Comment (albertz) [inline docs/index.rst]: I'm not sure I would put this to the very end. Rather, I think that should be almost as the very front, as this is one of the most important page for newcomers, to understand how RETURNN works (after maybe reading the basic usage page).

File: docs/index.rst [hunk 7]
```diff
@@ -49,32 +49,53 @@ Questions can also be asked on
...
+    internals/tf_lstm_benchmark.rst
```
Comment (albertz) [inline docs/index.rst]: I don't know if fits under "internal".

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 8]
```diff
@@ -0,0 +1,54 @@
+.. _recurrent_subnet:
+
+======================
+Recurrent Sub-Networks
+======================
+
+For many task it will be necessary to define multiple that are applied as recurrent network over a sequential input,
+especially when running a search over sequences. While basic recurrent layers such as LSTM variants are defined by using
```
Comment (albertz) [inline docs/getting_started/basic_usage/recurrent_subnet.rst]: Line breaks are again off here.

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+the "``rec``" layer and selecting the desired "``unit``", custom sub-networks can be defined by passing a network
+dictionary for the "``unit``" attribute.  The defined structure will then be applied for each position of the sequence.
+As for the global network, an "``output``" layer is required to define which values will be the output of the subnet.
+The layer outputs of the previous timesteps can be accessed by adding the prefix "prev:" to the layer names. Static data
+from outside the subnet can be accessed via the layer prefix "base:".
```
Comment (albertz) [inline docs/getting_started/basic_usage/recurrent_subnet.rst]: Some example units should be listed (here, or maybe below), e.g. `"NativeLstm2"` etc. And a link to the LSTM benchmark, such that the user can about relevant LSTM units.

Comment (albertz) [reply docs/getting_started/basic_usage/recurrent_subnet.rst]: Or maybe not here at all, but somewhere else.

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 2 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst]: What about forward mode? For me this was never an issue, because i need if cases to distinguish between training and testing anyways, and eval would never be set during training. I never had any problems with using eval, neither for forward nor for search, and i did not even know about the search_data option.

But anyways, this exactly the reason why i started writing the documentation, as many things seem to be unclear for me.

File: docs/generate_units.py (hunk 1 above)
Comment (JackTemaki) [reply docs/generate_units.py]: ok, then i will change the alias and the other script accordingly

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 2 above)
Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst]: Well, I intentionally left that case out of the documentation because it was not quite well-defined and the behavior could potentially change. (I would change it, but I'm not sure if that breaks someones setup...)

Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst]: From what I see in the code, following happens:
 * during `task='train'`, it will call `eval_model()` which evaluates on `dev` and `eval`, as `get_eval_dataset()` returns both. 
 * the same holds for `task='eval'`, unless `output_per_seq_file` is specified, then it asserts that only either (dev or eval, but not both) are set.
 * `task='forward'` only uses `eval` , nothing else is used
 * `task='search` uses `eval` as default,  but it can use `train` or `dev` if `seach_data` is set to `'train'` or `'dev'`. If `search_data` is not a string, it will interpret this as dataset.

You are definitely right, that this should be changed. So what of this should be part of the documentation?

Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst]: I'm not sure about changing the logic too much, as I don't know who actually depends on this behavior. But also, that is why I would not document `eval`, because the behavior is somewhat inconsistent, and new users should better just not use it at all. For the `search` task, people could just use `seach_data` (that's what I recommend anyway). We could independently document that for the search task. The forward task is anyway not documented in detail yet (and probably anyway rarely used; there are other things which are probably more important). And if needed, we could introduce a new key `forward_data` or so for that case.

File: docs/generate_units.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+Common units are:
+
+   * BasicLSTM (the cell), via official TF, pure TF implementation
+   * LSTMBlock (the cell), via tf.contrib.rnn.
+   * LSTMBlockFused, via tf.contrib.rnn. should be much faster than BasicLSTM
+   * CudnnLSTM, via tf.contrib.cudnn_rnn. This is experimental yet.
+   * NativeLSTM, our own native LSTM. should be faster than LSTMBlockFused.
+   * NativeLstm2, improved own native LSTM, should be the fastest and most powerful
+
+Note that the native implementations can not be in a recurrent subnetwork, as they process the whole sequence at once.
```
Comment (albertz) [inline docs/generate_units.py]: You should put a link to the benchmark here for the speed comparison.

File: docs/internals/tech_overview.rst [hunk 11]
```diff
@@ -58,69 +58,6 @@ Execution guide
...
-
-.. code-block:: python
-
-    network = {
-        "fw1": {"class": "linear", "activation": "relu", "dropout": 0.1, "n_out": 500},
-        "fw2": {"class": "linear", "activation": "relu", "dropout": 0.1, "n_out": 500, "from": ["fw1"]},
-        "output": {"class": "softmax", "loss": "ce", "from": ["fw2"]}
-    }
-
-The ``"class"`` key will get extracted from the layer arguments and the specific layer class will be used.
```
Comment (albertz) [inline docs/internals/tech_overview.rst]: Where is this text now? This is important for understanding how the network construction works.

File: docs/internals/tech_overview.rst [hunk 12]
```diff
@@ -58,69 +58,6 @@ Execution guide
...
-For Theano, the base layer class is :py:class:`NetworkBaseLayer.Container` and :py:class:`NetworkBaseLayer.Layer`;
-for TensorFlow, it is :py:class:`TFNetworkLayer.LayerBase`.
-E.g. that would use the :py:class:`TFNetworkLayer.LinearLayer` class,
-and the ``LinearLayer.__init__`` will accepts arguments like ``activation``.
-In the given example, all the remaining arguments will get handled by the base layer.
-
-The construction itself can be found for TensorFlow in :py:func:`TFNetwork.TFNetwork.construct_from_dict`,
-which starts from the output layers goes over the sources of a layer, which are defined by ``"from"``.
-If a layer does not define ``"from"``, it will automatically get the input from the dataset data.
```
Comment (albertz) [inline docs/internals/tech_overview.rst]: Where is this now?

File: docs/internals/tech_overview.rst [hunk 13]
```diff
@@ -58,69 +58,6 @@ Execution guide
...
-
-    "lstm2_fw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": 1, "from" : ["lstm1_fw", "lstm1_bw"] },
-    "lstm2_bw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": -1, "from" : ["lstm1_fw", "lstm1_bw"] },
-
-    "output" :   { "class" : "softmax", "loss" : "ce", "from" : ["lstm2_fw", "lstm2_bw"] }
-    }
-
-
-
-.. _tech_engine_train:
```
Comment (albertz) [inline docs/internals/tech_overview.rst]: Where is the `_tech_engine_train` tag now?

File: docs/internals/tech_overview.rst [hunk 14]
```diff
@@ -58,69 +58,6 @@ Execution guide
 - Then, depending on the ``task`` option, it might start ``engine.train``, ``engine.forward`` etc.
   (:py:func:`Engine.Engine.train` or :py:func:`TFEngine.Engine.train`), :ref:`tech_engine_train`.
 
-
-.. _tech_net_construct:
```
Comment (albertz) [inline docs/internals/tech_overview.rst]: Where is the `_tech_net_construct` tag now?

File: docs/layer_reference/index.rst [hunk 15]
```diff
@@ -1,93 +1,8 @@
...
-    * ``"glorot_uniform"``
-    * ``"orthogonal"``
-
-  * providing a dictionary for the initializer classes:
-
-    * Example: :code:`"forward_weights_init": {'class': 'VarianceScaling', 'scale': 0.5, 'mode': 'fan_out'}`
-
-The initialization is performed in :func:`TFUtil.get_initializer`.
-
-*Note:* the initalizers can be accessed both as e.g. ``"glorot_normal"`` or ``"glorot_normal_initializer"``.
```
Comment (albertz) [inline docs/layer_reference/index.rst]: All of this was removed now?

File: docs/layer_reference/index.rst [hunk 16]
```diff
@@ -1,93 +1,8 @@
...
-Connecting Layers
------------------
-
-In most cases it is sufficient to just specify a list of layer names for the **from** attribute. When no input is specified,
-it will automatically fallback to ``"data"``, which is the default input-data of the provided dataset. Depending on the
-definition of the ``feature`` and ``target`` keys (see :class:`Dataset.DatasetSeq`), the data can be accessed
-via ``from["data:DATA_KEY"]``. When specifying layers inside a recurrent unit (see :ref:`recurrent_layers`), two additional
-input prefixes are available, ``base`` and ``prev``. When trying to access layers from outside the recurrent unit, the prefix
-``base`` as to be used. Otherwise, only other layers inside the recurrent unit are recognised. ``prev`` can be used to access
-the layer output from the previous recurrent step (e.g. for target embedding feedback).
```
Comment (albertz) [inline docs/layer_reference/index.rst]: This was removed now?

**Comment by albertz** [comment]
Btw, some of the tests fail. Maybe just rebasing will fix some of them.
However, I think [this one](https://travis-ci.org/github/rwth-i6/returnn/jobs/662062349?utm_medium=notification&utm_source=github_status) is due to your changes.

File: docs/internals/tech_overview.rst (hunk 11 above)
Comment (JackTemaki) [reply docs/internals/tech_overview.rst]: I did not remove any text at all, this is part of the new files. This also holds for the next 4 comments. The tech overview should be changed, as the Theano information is more confusing than helping (it is deprecated anyways). Information on how to use RETURNN should all be in "basic_usage", and the information here should be organized in a different way. I also do not understand the name "technological overview", because this sounds like a description of available features of RETURNN, not like a list of internal classes. I will continue working on this when i have the time.

Comment (albertz) [reply docs/internals/tech_overview.rst]: But this tech overview was mostly about TF. It explains that it extracts the `class` key, calls `construct_from_dict`, and all this stuff. I don't see that this is part of the new files. Where exactly? In my opinion, this is some of the most important part of the documentation.

Comment (albertz) [reply docs/internals/tech_overview.rst]: Ah sorry, now I see it.

File: docs/getting_started/basic_usage/network.rst [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+The ``"class"`` key will get extracted from the layer arguments and the specific layer class will be used.
+For Theano, the base layer class is :py:class:`NetworkBaseLayer.Container` and :py:class:`NetworkBaseLayer.Layer`;
+for TensorFlow, it is :py:class:`TFNetworkLayer.LayerBase`.
+E.g. that would use the :py:class:`TFNetworkLayer.LinearLayer` class,
+and the ``LinearLayer.__init__`` will accepts arguments like ``activation``.
+In the given example, all the remaining arguments will get handled by the base layer.
+
+The construction itself can be found for TensorFlow in :py:func:`TFNetwork.TFNetwork.construct_from_dict`,
```
Comment (albertz) [inline docs/getting_started/basic_usage/network.rst]: Now you mix the usage and the technical implementation here. I'm not sure if that is a good idea.

File: docs/getting_started/basic_usage/network.rst [hunk 18]
```diff
@@ -0,0 +1,143 @@
+.. _network:
+
+=================
+Network Structure
+=================
+
+Construction
+------------------------------
```
Comment (albertz) [inline docs/getting_started/basic_usage/network.rst]: The length of `-` does not match here.

File: docs/getting_started/basic_usage/network.rst [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    "lstm2_bw" : { "class": "rec", "unit": "lstm", "n_out" : 500, "dropout": 0.1, "L2": 0.01, "direction": -1, "from" : ["lstm1_fw", "lstm1_bw"] },
+
+    "output" :   { "class" : "softmax", "loss" : "ce", "from" : ["lstm2_fw", "lstm2_bw"] }
+    }
+
+
+Defining Layers
+---------------
+
+Every usable layer with the tensorflow backend inherits from :class:`TFNetworkLayer.LayerBase`.
```
Comment (albertz) [inline docs/getting_started/basic_usage/network.rst]: It should be `TensorFlow`.

File: docs/getting_started/basic_usage/network.rst (hunk 17 above)
Comment (albertz) [reply docs/getting_started/basic_usage/network.rst]: Btw, technological overview was supposed to give an overview over the technology, i.e. an overview over the implementation. There is certainly an overlap of such a technological overview and a usage overview, but it's not the same thing. I would say that an overview of the implementation is also useful to better understand how to use it, although it's not strictly necessary. It's similar of e.g. knowing how Git works (the object model etc) and how to use the tool.

I still prefer to keep those separate. Although it's totally ok if there is some overlap and some parts are redundant. E.g. to describe the implementation, it is certainly useful to do that given a simple network example.

**Comment by albertz** [comment]
Are you still working on this? Can you maybe address my remaining comments?

**Comment by JackTemaki** [comment]
I still did not resolve all Issues yet. I had to prioritize other work, so I might finish this in may.

File: docs/getting_started/basic_usage/network.rst (hunk 17 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/network.rst:39]: I agree that this needs some reworking, also by changing/extending the "technological overview" part. The question is if this rework should still be part of this pull request, or if it can be included in the upcoming documentation updates.

File: docs/index.rst (hunk 7 above)
Comment (JackTemaki) [reply docs/index.rst:99]: I did not find a good place for it yet, maybe adding a "features" subsection under "getting started" that has a list of what RETURNN can do (or not do) would be an appropriate place.

File: docs/index.rst (hunk 6 above)
Comment (JackTemaki) [reply docs/index.rst:98]: I am not sure I can agree on this one, as for me this page mostly presents Information that is needed when you start making code changes to RETURNN, rather than using RETURNN.

**Comment by JackTemaki** [comment]
I think it would be good if we do not put too much effort in discussing the details yet, and wait for the responses/feedback of other people what they would like to have in the documentation, and where it should be. I mean this is only a starting point so far.

File: docs/index.rst (hunk 6 above)
Comment (albertz) [reply docs/index.rst:98]: But I would argue that you should have a good overview of how RETURNN works internally to really work with RETURNN. A lot of questions and problems just arise because people never really understood how RETURNN is designed.

File: docs/getting_started/basic_usage/network.rst (hunk 17 above)
Comment (albertz) [reply docs/getting_started/basic_usage/network.rst:39]: The technological overview was supposed to give an overview of how it works for newcomers. Really only an overview, not going too much into detail.

File: docs/index.rst (hunk 7 above)
Comment (albertz) [reply docs/index.rst:99]: I would move the technological overview to the `getting_started` directory.

And I would move the LSTM benchmark also somewhere else. Maybe `details`? Or `extended_usage`? This is really supposed to be for users, to give them an idea of what LSTM cell/unit they should use.

Comment (JackTemaki) [reply docs/index.rst:101]: Okay I put this to the beginning now, the network part is still redundant, but I will keep it for now until further modifications of the overview. For the LSTM benchmark I was still not sure, so I just put it at the end of getting started.

File: docs/layer_reference/utility.rst [hunk 20]
```diff
@@ -32,13 +32,6 @@ Print Layer
     :members:
     :undoc-members:
 
-Scale Gradient Layer
---------------------
-
-.. autoclass:: TFNetworkLayer.ScaleGradientLayer
-    :members:
-    :undoc-members:
-
```
Comment (albertz) [inline docs/layer_reference/utility.rst:41]: Why was this removed?

File: docs/index.rst (hunk 6 above)
Comment (albertz) [reply docs/index.rst:100]: Is it correct that `tech_overview` is both under `internals` and `getting_started` now?

File: docs/index.rst [hunk 21]
```diff
@@ -49,32 +49,56 @@ Questions can also be asked on
...
```
Comment (albertz) [inline docs/index.rst:101]: `internals` is not correct anymore, right?

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 22]
```diff
@@ -115,6 +83,27 @@ network
...
+    The total number of frames. A mini-batch has at least a time-dimension
+    and a batch-dimension (or sequence-dimension), and depending on dense or sparse,
+    also a feature-dimension.
+    ``batch_size`` is the upper limit for ``time * sequences`` during creation of the mini-batches.
+
+max_seqs
+    The maximum number of sequences in one mini-batch.
+
+..
+    chunking
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:100]: Is that on purpose here, the `..`, and then `chunking` indented?

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 23]
```diff
@@ -134,69 +125,16 @@ log_verbosity
     An integer. Common values are 3 or 4. Starting with 5, you will get an output per mini-batch.
 
 
-There are much more params, and more details to many of the listed ones.
-See the code for more details.
-All config params can also be passed as command line params.
-See the code for some usage. The generic form is ``++param value``.
+There are much more params, and more details to many of the listed ones. Details on the parameters can be found in the
+:ref:`parameter reference <parameter_reference>`. As the reference is still incomplete, please watch out for additional
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:129]: There should be a newline after the dot (`.`), such that diffs/changes are nicer and more structured.

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 24]
```diff
@@ -134,69 +125,16 @@ log_verbosity
...
+parameters that can be found in the code. All config params can also be passed as command line params.
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:130]: Newline after dot.

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 25]
```diff
@@ -134,69 +125,16 @@ log_verbosity
...
+The generic form is ``++param value``, but more options are available. Please See the code for some usage.
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:131]: Newline after dot.

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 26]
```diff
@@ -0,0 +1,54 @@
+.. _recurrent_subnet:
+
+======================
+Recurrent Sub-Networks
+======================
+
+For many task it will be necessary to define multiple that are applied as recurrent network over a sequential input,
+especially when running a search over sequences. While basic recurrent layers such as LSTM variants are defined by using
+the "``rec``" layer and selecting the desired "``unit``", custom sub-networks can be defined by passing a network
+dictionary for the "``unit``" attribute.  The defined structure will then be applied for each position of the sequence.
```
Comment (albertz) [inline docs/getting_started/basic_usage/recurrent_subnet.rst:10]: There should be a newline after the dot.

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+            "output": {
+              "class": "linear",
+              "from": ["prev:output", "data:source"],
+              "activation": "relu",
+              "n_out": n_out},
+          },
+          "n_out": n_out,
+      }
+
+Layers with recurrent dependencies and hidden states (e.g. LSTMs) can be added as "``rnn_cell``" layer. For available
```
Comment (albertz) [inline docs/getting_started/basic_usage/recurrent_subnet.rst:33]: There should be a newline after the dot.

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+              "energy_tanh": {"class": "activation", "activation": "tanh", "from": ["energy_in"]},
+              "energy": {"class": "linear", "activation": None, "with_bias": False, "from": ["energy_tanh"], "n_out": 128},
+              "att_weights": {"class": "softmax_over_spatial", "from": ["energy"]},  # (B, enc-T, H)
+              "att": {"class": "generic_attention", "weights": "att_weights", "base": "base:enc_value"},  # (B, H, V)
+              "output": {"class": "rnn_cell", "unit": "LSTMBlock", "from": ["prev:att"], "n_out": 256, 'target': 'data'},  # transform
+          }
+          "n_out": n_out
+      }
+
+The ``from`` attribute can be empty when the output as a target. The sequence length will then be determined by this target.
```
Comment (albertz) [inline docs/getting_started/basic_usage/recurrent_subnet.rst:54]: There should be a newline after the dot.

File: docs/layer_reference/utility.rst (hunk 20 above)
Comment (JackTemaki) [reply docs/layer_reference/utility.rst:41]: I removed this because it caused an error, I did not notice this was a typo only.

Comment (albertz) [reply docs/layer_reference/utility.rst:41]: Ok, can you maybe put it in again then?

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 22 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:100]: Yes, this is the multiline comment function for Sphinx. Although batching and chunking were in the original file, I think that chunking is a very special parameter and might not be appropriate under basic usage. For batching, I excluded it because I am still not sure if people should rather be encouraged to use "seq_ordering" in the datasets explicitly.

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 23 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:129]: I now went through all files and changed it to at most one sentence per line

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 22 above)
Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst:100]: `chunking` is very basic. Everyone working with hybrid models would want to use that. This definitely should be in there.

Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:100]: I would disagree here, I think at this point only parameters that are needed for ALL models should be mentioned, as otherwise this list needs to be very long. If you say the whole documentation should be focused on ASR, as this is currently its main purpose, then yes it can stay.

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 29]
```diff
@@ -27,58 +27,26 @@ as well as some training parameters.
 Some common config parameters:
 
 task
-    The task, such as ``train`` or ``forward``.
+    The task, such as ``train``, ``forward`` or ``search``.
 
 device
     E.g. ``gpu`` or ``cpu``. Can also be ``gpu0,gpu1`` for multi-GPU training.
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:33]: There should also be a newline.
And better or maybe just remove the comment for multi-GPU training. This is for Theano.

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 22 above)
Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst:100]: I would not say focused, but I would say that ASR (and also hybrid models) are still something very important, so I would definitely leave this in for now. Also, this is more generic for all kinds of frame-wise sequence trainings, not just ASR.

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 30]
```diff
@@ -115,6 +83,27 @@ network
     e.g. you can choose between different LSTM implementations, or GRU, or standard RNN, etc.
     See :class:`TFNetworkRecLayer.RecLayer` or :class:`NetworkRecurrentLayer.RecurrentUnitLayer`.
     See also :ref:`tf_lstm_benchmark`.
+..
+    batching
+        The sorting variant when the mini-batches are created. E.g. ``random``.
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:88]: Either put it back in, or completely remove.
I don't think that we should have commented out text in the documentation files.
If you really want to remove that, the `seq_ordering` option for the datasets should be emphasized more (which is completely missing currently).

File: docs/index.rst (hunk 6 above)
Comment (albertz) [reply docs/index.rst:100]: Ping here. (Maybe you missed it?)

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 30 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:88]: The `seq_ordering` option is mentioned in the full parameter overview (in the entry for `batching`). This is also the reason why I removed `chunking` here, the same description is available under configuration reference / training. I also have a sentence recommending `seq_ordering` over `batching`, but this is not part of this PR yet.

Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:88]: Maybe to not cause confusion: at this location (basic usage), I only wanted to state the most general parameters, which are needed for every model. Everything else is sorted into the parameter categories for better visibility.

Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst:88]: But I would argue that it is pretty basic to have sequence shuffling.
But anyway, there should not be commented out text.

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 29 above)
Comment (albertz) [reply docs/getting_started/basic_usage/basic_usage.rst:33]: Note that this is an option not necessarily needed. It would automatically detect the device. By explicitly defining it, you just enforce it (and it would raise an exception if it cannot get what you requested).

File: docs/getting_started/basic_usage/basic_usage.rst [hunk 31]
```diff
@@ -27,58 +27,26 @@ as well as some training parameters.
...
+train / dev / eval
     The datasets. This can be a filename to a hdf-file.
     Or it can be a dict with an entry ``class`` where you can choose a from a variety
     of other dataset implementations, including many synthetic generated data.
+    ``train`` and ``dev`` are used during training, while ``eval`` is usually used to define the dataset for the
+    ``forward`` or ``search`` task.
 
-extern_data (former num_outputs)
+extern_data
+    A dictionary
```
Comment (albertz) [inline docs/getting_started/basic_usage/basic_usage.rst:46]: Shouldn't there be a dot at the end?

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 30 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:88]: Ok. If your are okay with it i would leave batching in (with the seq_ordering note), and chunking out, as there is a link to the other parameters below.

File: docs/getting_started/basic_usage/basic_usage.rst (hunk 31 above)
Comment (JackTemaki) [reply docs/getting_started/basic_usage/basic_usage.rst:46]: This seems to be wrong in general, this text is incomplete

**Comment by JackTemaki** [comment]
I understand that there are still many things left to be corrected, and I will also continue to work on the documentation until I have the feeling it is somewhat complete.

File: docs/getting_started/basic_usage/data.rst [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+
+    train = {'class': 'HDFDataset', 'files': ['path/to/training_data.hdf']}
+
+Most datasets follow the convention that the input data is sequential and has the label "``data``", and the target data
+is sparse and has the label "``classes``".
+In the case of the hdf file it could be that the input data are 100-dimensional MFCCs
+and the target data are 5,000 word classes.
+Then, ``extern_data`` would have to be defined as:
+
+.. code-block::
```
Comment (patrick-wilken) [inline docs/getting_started/basic_usage/data.rst:26]: This block is not showing on readthedocs. Because of missing "python"? (I have no clue)

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 33]
```diff
@@ -0,0 +1,56 @@
+.. _recurrent_subnet:
+
+======================
+Recurrent Sub-Networks
+======================
+
+For many task it will be necessary to define multiple that are applied as recurrent network over a sequential input,
```
Comment (patrick-wilken) [inline docs/getting_started/basic_usage/recurrent_subnet.rst:7]: multiple... layers?

File: docs/getting_started/basic_usage/recurrent_subnet.rst [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+              "energy_tanh": {"class": "activation", "activation": "tanh", "from": ["energy_in"]},
+              "energy": {"class": "linear", "activation": None, "with_bias": False, "from": ["energy_tanh"], "n_out": 128},
+              "att_weights": {"class": "softmax_over_spatial", "from": ["energy"]},  # (B, enc-T, H)
+              "att": {"class": "generic_attention", "weights": "att_weights", "base": "base:enc_value"},  # (B, H, V)
+              "output": {"class": "rnn_cell", "unit": "LSTMBlock", "from": ["prev:att"], "n_out": 256, 'target': 'data'},  # transform
+          }
+          "n_out": n_out
+      }
+
+The ``from`` attribute can be empty when the output as a target.
```
Comment (patrick-wilken) [inline docs/getting_started/basic_usage/recurrent_subnet.rst:55]: "when using"?

**Comment by patrick-wilken** [comment]
Thanks for working on the docs! I commented on a few things I found skimming through it.

**Comment by albertz** [comment]
I noticed that quite a few links got broken. E.g. [multi GPU training](https://returnn.readthedocs.io/en/latest/multi_gpu.html) or [TF LSTM benchmark](https://returnn.readthedocs.io/en/latest/tf_lstm_benchmark.html), which was both linked on multiple other sites.
Is there a way to redirect those?

**Comment by JackTemaki** [comment]
According to the RTD docs, this is possible, but this has to be done in the RTD project, I can not do it myself without access.
The guide is here:
https://docs.readthedocs.io/en/latest/user-defined-redirects.html

**Comment by albertz** [comment]
The project home on readthedocs is [here](https://readthedocs.org/projects/returnn/).
If you tell me your username (on readthedocs), I can add you as a maintainer.

I actually was thinking just about a dummy page with a content like "this site was moved <here>". Maybe with some JS code (if possible) for an automatic redirect.
But I did the redirect via this official feature now.
It seems to work for multi GPU but not for the TF LSTM benchmark for some reason. Maybe I just need to wait. Or maybe I did some mistake / typo? Although it all looks correct to me.

image

File: docs/getting_started/basic_usage/data.rst (hunk 32 above)
Comment (albertz) [reply docs/getting_started/basic_usage/data.rst:26]: You can always check the build logs ([here](https://readthedocs.org/projects/returnn/builds/); you also find the link on the doc itself).
(But I don't see anything, except some (probably unrelated) Theano errors.)
